bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -86,44 +84,44 @@
      */
     private class ClientCallbacks implements MqttCallback {
         @Override
-        public synchronized void connectionLost(Throwable t) {
-            if (t instanceof MqttException) {
-                MqttException e = (MqttException) t;
-                logger.info("MQTT connection to '{}' was lost: {} : ReasonCode {} : Cause : {}", getName(),
-                        e.getMessage(), e.getReasonCode(),
-                        (e.getCause() == null ? "Unknown" : e.getCause().getMessage()));
+        public synchronized void connectionLost(@Nullable Throwable exception) {
+            assert exception != null;
```

Comment:
`exception` being null seems to be a legit case. Its way too hard to throw errors here.
We do not use `assert` in ESH at all: Assertions need to be switched on during compilation what we don't do. Thus they will be removed from the resulting class files. And in this particular case it is hiding a NPE when `exception` actually _is_ null.
Please review other occurrences and replace them by throwing IllegalArgumentException or IllegalStateException.

Reply:
I checked with the documentation. All parameters on this interface are meant to be non null. The annotations are just missing and assert is basically used to compensate. I have added some comments

Reply:
> Assertions need to be switched on during compilation what we don't do.

That's not correct.
assertions could be enabled at runtime to use additional checks (e.g. in the development phase) that are not enabled default

Reply:
Ah I see (http://etutorials.org/Programming/Java+performance+tuning/Chapter+6.+Exceptions+Assertions+Casts+and+Variables/6.2+Assertions/). So it actually imposes a runtime overhead in contrast to C/C++. I'll use local variables then to get rid of the warnings.